### PR TITLE
Add GroupNorm

### PR DESCRIPTION
**Pull request by zjlablichenyang**
Add GroupNorm Module
GroupNorm

GroupNorm Doctest
test_groupnorm

### CODE REVIEW

File: oneflow/python/nn/modules/groupnorm.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+        normalized = flow.experimental.reshape(normalized, shape=[origin_shape[0],  self.num_channels, -1])
+        normalized = self.weight * normalized + self.bias
+        res= flow.experimental.reshape_like(normalized, like=input)
+
+        return res
+
+if __name__ == "__main__":
+    import doctest
+
+    doctest.testmod()
```
Comment (doombeaker) [inline oneflow/python/nn/modules/groupnorm.py:117]: 文末换行

File: oneflow/python/nn/modules/groupnorm.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+limitations under the License.
+"""
+
+import oneflow as flow
+from oneflow.python.framework.tensor import Tensor
+from oneflow.python.oneflow_export import oneflow_export, experimental_api
+from oneflow.python.nn.module import Module
+
+
+@oneflow_export("nn.GroupNorm")
```
Comment (doombeaker) [inline oneflow/python/nn/modules/groupnorm.py:23]: 需要在 `oneflow/python/test/modules` 目录添加对应的 `test_xxx` 文件，测试

File: oneflow/python/nn/modules/groupnorm.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+@experimental_api
+class GroupNorm(Module):
+    r"""The interface is consistent with PyTorch.
+    The documentation is referenced from:
+    https://pytorch.org/docs/stable/generated/torch.nn.GroupNorm.html
+
+    Applies Group Normalization over a mini-batch of inputs as described in
```
Comment (doombeaker) [inline oneflow/python/nn/modules/groupnorm.py:30]: 需要编译并截图 docstring

File: oneflow/python/nn/modules/groupnorm.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    the paper `Group Normalization <https://arxiv.org/abs/1803.08494>`__
```
Comment (doombeaker) [inline oneflow/python/nn/modules/groupnorm.py:31]: 这种超链接形式可能在我们的编译过程中不能很好支持，请留意编译后的结果，并且确保编译过程中没有 warning（否则现在也通过不了CI）

File: oneflow/python/nn/modules/groupnorm.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+        if self.affine:
+            flow.nn.init.ones_(self.weight)
+            flow.nn.init.zeros_(self.bias)
+
+    def forward(self, input:Tensor) -> Tensor:
+        assert len(input.shape) >= 3
+        assert (input.shape[1] == self.num_channels), "The channels of input tensor must equal num_channels"
+        origin_shape = input.shape
+        reshape_to_1d = flow.experimental.reshape(input, shape=[origin_shape[0], self.num_groups, -1])
+        (mean, variance) = flow.experimental.nn.moments(reshape_to_1d, [2], keepdims=True)
```
Comment (doombeaker) [inline oneflow/python/nn/modules/groupnorm.py:106]: pytorch 没有 `nn.momentd`， OneFlow 也没有迁移 `flow.experimental.nn.moments`。
这里可以用已经迁移的 `flow.experimental.mean` 和还没有迁移的 `flow.experimental.var`（但是已经有对应的老接口 `flow.reduce_variance` 分别算出来。不使用 `nn.moments`。

所以这个PR里，建议同时把 `flow.experimental.var` 也给对齐、搬运了（这也是为什么我说真的需求会在搬运过程中发现）。

PS：像 `flow.experimental.nn.moments` 其实不存在这种问题，自己试一试就会发现，所以搬运过程中还是要把自己力所能及的test给跑通比较好，自己就会发现问题，这样可以省掉来回沟通的时间成本

File: oneflow/python/test/modules/test_groupnorm.py [hunk 6]
```diff
@@ -0,0 +1,360 @@
...
+        dtype=np.float32,
+    )
+
+    x = flow.Tensor(input_arr, device=flow.device(device))
+    m = flow.nn.GroupNorm(num_groups=1, num_channels=2).to(device=flow.device(device))
+    y = m(x)
+    test_case.assertTrue(np.allclose(y.numpy(), output, 1e-5, 1e-5))
+
+
+def _test_groupnorm_v2(test_case, device):
```
Comment (doombeaker) [inline oneflow/python/test/modules/test_groupnorm.py:106]: 这个名字可以用更有含义的代替么 `v2` 的信息太少了，比如 `smaller_shape` 什么的

File: oneflow/python/nn/modules/groupnorm.py [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+
+        self.reset_parameters()
+
+    def reset_parameters(self) -> None:
+        if self.affine:
+            flow.nn.init.ones_(self.weight)
+            flow.nn.init.zeros_(self.bias)
+
+    def forward(self, input: Tensor) -> Tensor:
+        assert len(input.shape) >= 3
```
Comment (Flowingsun007) [inline oneflow/python/nn/modules/groupnorm.py:105]: assert后需要加一下提示信息

File: oneflow/python/nn/modules/groupnorm.py [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+    Args:
+        num_groups (int): number of groups to separate the channels into
+        num_channels (int): number of channels expected in input
+        eps: a value added to the denominator for numerical stability. Default: 1e-5
+        affine: a boolean value that when set to ``True``, this module
+            has learnable per-channel affine parameters initialized to ones (for weights)
+            and zeros (for biases). Default: ``True``.
+
+    Shape:
+        - Input: :math:`(N, C, *)` where :math:`C=\text{num_channels}`
```
Comment (BBuf) [inline oneflow/python/nn/modules/groupnorm.py:57]: 这个公式显示好像不正常

File: oneflow/python/nn/modules/groupnorm.py [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+
+        >>> input = flow.Tensor(np.random.randn(20, 6, 10, 10))
+        >>> # Separate 6 channels into 3 groups
+        >>> m = flow.nn.GroupNorm(3, 6)
+        >>> # Separate 6 channels into 6 groups (equivalent with InstanceNorm)
+        >>> m = flow.nn.GroupNorm(6, 6)
+        >>> # Put all 6 channels into a single group (equivalent with LayerNorm)
+        >>> m = flow.nn.GroupNorm(1, 6)
+        >>> # Activating the module
+        >>> output = m(input)
```
Comment (BBuf) [inline oneflow/python/nn/modules/groupnorm.py:76]: 看一下每种情况的输出shape

Comment (zjlablichenyang) [reply oneflow/python/nn/modules/groupnorm.py:76]: 每种情况输出的shape都一样

File: oneflow/python/nn/modules/groupnorm.py (hunk 8 above)
Comment (zjlablichenyang) [reply oneflow/python/nn/modules/groupnorm.py:57]: 这里已经修改过了，我更新一下截图

File: oneflow/python/test/modules/test_groupnorm.py [hunk 10]
```diff
@@ -0,0 +1,360 @@
...
+import unittest
+from collections import OrderedDict
+
+import numpy as np
+
+import oneflow.experimental as flow
+from test_util import GenArgList
+
+
+input_arr = np.array(
```
Comment (BBuf) [inline oneflow/python/test/modules/test_groupnorm.py:25]: 这个是多余的？

File: oneflow/python/nn/modules/groupnorm.py (hunk 9 above)
Comment (BBuf) [reply oneflow/python/nn/modules/groupnorm.py:76]: 好吧

File: oneflow/python/test/modules/test_groupnorm.py (hunk 10 above)
Comment (zjlablichenyang) [reply oneflow/python/test/modules/test_groupnorm.py:25]: 是的，我去删掉

File: oneflow/python/test/modules/test_groupnorm.py [hunk 11]
```diff
@@ -0,0 +1,360 @@
...
+                [
+                    [-0.8472, 1.3012, -1.1065, 0.9348],
+                    [1.0346, 1.5703, 0.2419, -0.7048],
+                    [0.6957, -0.4523, -0.8819, 1.0164],
+                ],
+            ],
+        ],
+        dtype=np.float32,
+    )
+    output = np.array(
```
Comment (BBuf) [inline oneflow/python/test/modules/test_groupnorm.py:70]: 感觉前向的输出可以用numpy模拟一下？

Comment (zjlablichenyang) [reply oneflow/python/test/modules/test_groupnorm.py:70]: 别的instancenorm，batchnorm都是这样直接写的

File: oneflow/python/nn/modules/groupnorm.py (hunk 7 above)
Comment (zjlablichenyang) [reply oneflow/python/nn/modules/groupnorm.py]: 已添加

**Review by BBuf — APPROVED** [review]
